source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,300 @@
+                        const BaseVerifierImpl* parent);
+
+// Base verifier for requires all or any.
+class BaseGroupVerifierImpl : public BaseVerifierImpl {
+public:
+  BaseGroupVerifierImpl(const BaseVerifierImpl* parent) : BaseVerifierImpl(parent) {}
+
+  void verify(ContextSharedPtr context) const override {
+    auto& ctximpl = static_cast<ContextImpl&>(*context);
+    for (const auto& it : verifiers_) {
+      if (ctximpl.getCompletionState(this).is_completed_) {
+        return;
+      }
+      it->verify(context);
+    }
+  }
+
+protected:
+  // The list of requirement verifiers
+  std::vector<VerifierPtr> verifiers_;
+};
+
+// Requires any verifier.
+class AnyVerifierImpl : public BaseGroupVerifierImpl {
+public:
+  AnyVerifierImpl(const JwtRequirementOrList& or_list, const AuthFactory& factory,
+                  const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                  const Extractor& extractor_for_allow_fail, const BaseVerifierImpl* parent)
+      : BaseGroupVerifierImpl(parent) {
+    for (const auto& it : or_list.requirements()) {
+      verifiers_.emplace_back(innerCreate(it, providers, factory, extractor_for_allow_fail, this));
+    }
+  }
+
+  void onComplete(const Status& status, ContextImpl& context) const override {
+    auto& completion_state = context.getCompletionState(this);
+    if (completion_state.is_completed_) {
+      return;
+    }
+    if (++completion_state.number_completed_children_ == verifiers_.size() ||
+        Status::Ok == status) {
+      completion_state.is_completed_ = true;
+      completeWithStatus(status, context);
+    }
+  }
+};
+
+// Requires all verifier
+class AllVerifierImpl : public BaseGroupVerifierImpl {
+public:
+  AllVerifierImpl(const JwtRequirementAndList& and_list, const AuthFactory& factory,
+                  const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                  const Extractor& extractor_for_allow_fail, const BaseVerifierImpl* parent)
+      : BaseGroupVerifierImpl(parent) {
+    for (const auto& it : and_list.requirements()) {
+      verifiers_.emplace_back(innerCreate(it, providers, factory, extractor_for_allow_fail, this));
+    }
+  }
+
+  void onComplete(const Status& status, ContextImpl& context) const override {
+    auto& completion_state = context.getCompletionState(this);
+    if (completion_state.is_completed_) {
+      return;
+    }
+    if (++completion_state.number_completed_children_ == verifiers_.size() ||
+        Status::Ok != status) {
+      completion_state.is_completed_ = true;
+      completeWithStatus(status, context);
+    }
+  }
+};
+
+// Match all, for requirement not set
+class AllowAllVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowAllVerifierImpl(const BaseVerifierImpl* parent) : BaseVerifierImpl(parent) {}
+
+  void verify(ContextSharedPtr context) const override {
+    onComplete(Status::Ok, static_cast<ContextImpl&>(*context));
```

Comment:
you can call completeWithStatus() directly here.  Not need to overload onComplete() function